File: leetcode-04/solutions/python3/0138.py
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        if head in self.dict:
            return self.dict[head]

        self.dict[head] = Node(head.val, None, None)
        self.dict[head].next = self.copyRandomList(head.next)
        self.dict[head].random = self.copyRandomList(head.random)

        return self.dict[head]

    dict = {}

File: leetcode-04/solutions/cpp/0138.cpp
class Solution {
 public:
  Node* copyRandomList(Node* head) {
    if (!head) return NULL;
    if (map.count(head)) return map[head];

    map[head] = new Node(head->val, NULL, NULL);
    map[head]->next = copyRandomList(head->next);
    map[head]->random = copyRandomList(head->random);

    return map[head];
  }

 private:
  unordered_map<Node*, Node*> map;
};